### PR TITLE
Fetching frontier tensors to frontend for ORTModule

### PR DESCRIPTION
**Pull request by ashbhandare**
Part [1/N]
This is the first part in adding support for the Gradient Accumulation optimization.
This PR adds support to fetch the names of the "frontier" tensors to the frontend.
The "frontier" tensors as described in the PR, are tensors that only depend on the value of the parameters of the model.
For eg, The output of a Cast of a parameter will only change when the parameter changes.
Initial implementation is targeting such "Cast" of parameters, which will be extended to find further downstream tensors that can be cached.

### CODE REVIEW

**Review by SherlockNoMad — DISMISSED** [review]

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.cc [hunk 1]
```diff
@@ -89,6 +89,10 @@ Status OrtModuleGraphBuilder::Build(const std::vector<std::vector<int64_t>>* inp
 
   ORT_RETURN_IF_ERROR(BuildGradientGraph(x_node_arg_names));
 
+  if (config_.enable_caching) {
+    GetFrontierTensors();
```
Comment (codemzs) [inline orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:93]: >GetFrontierTensors [](http://example.com/codeflow?start=4&length=18)

The graph gets transformed after this line of code in InferenceSession:Initialize(). Just want to make sure it is ok to fetch the frontier tensors here.

**Comment by codemzs** [comment]
Looks good except for that one [comment](https://github.com/microsoft/onnxruntime/pull/8086#discussion_r657464287).

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.cc (hunk 1 above)
Comment (ashbhandare) [reply orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:93]: With the current plan for fetching the frontier tensors- as graph outputs, we have to fetch the frontier tensors before we transform the graph at this point. This is okay for the initial implementation that will focus on Cast outputs.

Comment (codemzs) [reply orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:93]: Thanks, can you please put this in a comment at line 93?

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.cc [hunk 2]
```diff
@@ -206,6 +210,21 @@ Status OrtModuleGraphBuilder::BuildGradientGraph(const std::unordered_set<std::s
...
+void OrtModuleGraphBuilder::GetFrontierTensors() {
+  const Graph& graph = gradient_model_->MainGraph();
+  for (const auto& param : graph_info_.initializer_names_to_train) {
+    std::vector<const Node*> consumer_nodes = graph.GetConsumerNodes(param);
+    // Initial support is limited to caching Cast output. This can
+    // be extended to accomodate more ops whose result depends only
+    // on the weight tensor which is a WIP.
+    for (const Node* node : consumer_nodes) {
+      if (node != nullptr && node->OpType() == "Cast") {
+        graph_info_.frontier_node_arg_map[param] = node->OutputDefs()[0]->Name();
```
Comment (codemzs) [inline orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:222]: >graph_info_.frontier_node_arg_map[param] = node->OutputDefs()[0]->Name(); [](http://example.com/codeflow?start=8&length=73)

nit: might be good to assert that "cast" op has exactly one output.

Comment (ashbhandare) [reply orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:222]: it does not need to have just one output

**Review by codemzs — APPROVED** [review]
:shipit:

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.cc (hunk 1 above)
Comment (ashbhandare) [reply orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:93]: will do if there are more changes else in the next PR

**Review by baijumeswani — APPROVED** [review]

**Pull request merged**